File: src/main/python/AoC2016_05.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 5
#

from hashlib import md5
from aoc import my_aocd
from aoc.common import spinner, log


def _find_md5_starting_with_5_zeroes(input_: str, index: int) -> (int, int):
    val = input_
    while val[:5] != "00000":
        index += 1
        spinner(index)
        str2hash = input_ + str(index)
        val = md5(str2hash.encode()).hexdigest()  # nosec
    return val, index


def part_1(inputs: tuple[str]) -> str:
    assert len(inputs) == 1
    index = 0
    result = ""
    for i in range(8):
        log(i)
        val, index = _find_md5_starting_with_5_zeroes(inputs[0], index)
        result += val[5]
        log(result)
    return result


def part_2(inputs: tuple[str]) -> str:
    assert len(inputs) == 1
    index = 0
    result = list(" " * 8)
    seen = set()
    while True:
        val, index = _find_md5_starting_with_5_zeroes(inputs[0], index)
        temp = val[5]
        if temp not in "01234567" or temp in seen:
            continue
        seen.add(temp)
        result[int(temp)] = val[6]
        log(result)
        if result.count(' ') == 0:
            break
    return "".join(result)


TEST = "abc".splitlines()


def main() -> None:
    my_aocd.print_header(2016, 5)

    assert part_1(TEST) == "18f47a30"
    assert part_2(TEST) == "05ace8e3"

    inputs = my_aocd.get_input(2016, 5, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")


if __name__ == '__main__':
    main()

File: src/main/java/AoC2016_05.java
import com.github.pareronia.aoc.CharArrayUtils;
import com.github.pareronia.aoc.codec.MD5;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@SuppressWarnings({"PMD.NoPackage", "PMD.ClassNamingConventions"})
public final class AoC2016_05 extends SolutionBase<String, String, String> {

    private AoC2016_05(final boolean debug) {
        super(debug);
    }

    public static AoC2016_05 create() {
        return new AoC2016_05(false);
    }

    public static AoC2016_05 createDebug() {
        return new AoC2016_05(true);
    }

    @Override
    protected String parseInput(final List<String> inputs) {
        return inputs.get(0);
    }

    private ValueAndIndex findMd5StartingWith5Zeroes(final String input, final Integer index) {
        String val = "XXXXX";
        int i = index;
        while (!"00000".equals(val.substring(0, 5))) {
            i++;
            final String toHash = input + i;
            val = MD5.md5Hex(toHash);
        }
        return ValueAndIndex.of(val, i);
    }

    @Override
    public String solvePart1(final String input) {
        Integer index = 0;
        final StringBuilder result = new StringBuilder();
        for (int i = 0; i < 8; i++) {
            log(i);
            final ValueAndIndex md5 = findMd5StartingWith5Zeroes(input, index);
            index = md5.index;
            result.append(md5.value.charAt(5));
            log(result);
        }
        return result.toString();
    }

    @Override
    public String solvePart2(final String input) {
        final char[] validPositions = {'0', '1', '2', '3', '4', '5', '6', '7'};
        Integer index = 0;
        final char[] result = {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '};
        final Set<Character> seen = new HashSet<>();
        while (true) {
            final ValueAndIndex md5 = findMd5StartingWith5Zeroes(input, index);
            final String val = md5.value;
            index = md5.index;
            final char temp = val.charAt(5);
            if (!CharArrayUtils.contains(validPositions, temp) || seen.contains(temp)) {
                continue;
            }
            seen.add(temp);
            final Integer position = Integer.valueOf(String.valueOf(temp));
            result[position] = val.charAt(6);
            log(String.valueOf(result));
            if (!CharArrayUtils.contains(result, ' ')) {
                break;
            }
        }
        return String.valueOf(result);
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "18f47a30"),
        @Sample(method = "part2", input = TEST, expected = "05ace8e3"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST = "abc";

    record ValueAndIndex(String value, int index) {
        public static ValueAndIndex of(final String value, final int index) {
            return new ValueAndIndex(value, index);
        }
    }
}
